lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -253,23 +255,179 @@ struct BinaryOpLowering : public OpConversionPattern<BinOp> {
+              for (auto element : type.getElements()) {
+                convertedElements.push_back(
+                    {element.name, removeSignedness(element.type)});
+              }
+              return hw::StructType::get(type.getContext(), convertedElements);
+            })
+            .Default([](auto type) { return type; });
+
+    return convertedType;
+  }
+
+  // Returns true if any subtype in 'type' has signedness semantics.
+  bool hasSignednessSemantics(Type type) {
+    auto it = conversionCache.find(type);
+    if (it != conversionCache.end())
+      return it->second.hadSignednessSemantics;
+
+    auto match =
+        llvm::TypeSwitch<Type, bool>(type)
+            .Case<IntegerType>([](auto type) { return !type.isSignless(); })
+            .Case<hw::ArrayType>([this](auto type) {
+              return hasSignednessSemantics(type.getElementType());
+            })
+            .Case<hw::StructType>([this](auto type) {
+              return llvm::any_of(type.getElements(), [this](auto element) {
+                return this->hasSignednessSemantics(element.type);
+              });
+            })
+            .Default([](auto type) { return false; });
+
+    if (match) {
+      // Prime the conversion cache by pre-converting the type.
+      conversionCache[type] = {removeSignedness(type), true};
+    } else {
+      // Prime the conversion cache by not converting the type - this prevents
+      // iterating through the type on future removeSignedness calls for this
+      // type.
+      conversionCache[type] = {type, false};
+    }
+
+    return match;
+  }
+
+  bool hasSignednessSemantics(TypeRange types) {
+    return llvm::any_of(types,
+                        [this](Type t) { return hasSignednessSemantics(t); });
+  }
+
+private:
+  // Memoizations for signedness info and conversions.
+  struct ConvertedType {
+    Type type;
+    bool hadSignednessSemantics;
+  };
+  llvm::DenseMap<Type, ConvertedType> conversionCache;
+};
+
+HWArithToHWTypeConverter::HWArithToHWTypeConverter() {
+  // Pass any type through the signedness remover.
+  addConversion([this](Type type) { return removeSignedness(type); });
+}
+
+// Adds the ArgResOpConversion for 'TOp' to the set of conversion patterns, as
+// well as a legality check on the conversion status of the op's operands and
+// results.
+template <typename TOp>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext());
+  target.addDynamicallyLegalOp<TOp>([&](TOp op) {
+    return !typeConverter.hasSignednessSemantics(op->getOperandTypes()) &&
+           !typeConverter.hasSignednessSemantics(op->getResultTypes());
+  });
+}
+
+template <typename TOp, typename TOp2, typename... OpTs>
+static void addOperandConversion(ConversionTarget &target,
```

Comment:
No need for this helper if [fold expressions](https://en.cppreference.com/w/cpp/language/fold) are used.
I.e. the functions could be unified to:
```C++
template <typename... TOp>
static void addOperandConversion(ConversionTarget &target,
                                 RewritePatternSet &patterns,
                                 HWArithToHWTypeConverter &typeConverter) {
  (patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext()),
   ...);
  (target.addDynamicallyLegalOp<TOp>([&](TOp op) {
    return !typeConverter.hasSignednessSemantics(op->getOperandTypes()) &&
           !typeConverter.hasSignednessSemantics(op->getResultTypes());
  }),
   ...);
}
```
Though, I could understand if that is considered as too obscure for maintainability.

Reply:
Awesome - I like it, i like templates, and i like code reductions!